### PR TITLE
[TypeInfo] Introduce component

### PR DESCRIPTION
**Pull request by mtarld**

### CODE REVIEW

File: src/Symfony/Component/TypeInfo/Type.php [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+    public function isCollection(): bool
+    {
+        return $this->isCollection;
+    }
+
+    /**
+     * @throws LogicException
+     */
+    public function getCollectionKeyType(): self
```
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: Remove

File: src/Symfony/Component/TypeInfo/Type.php [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    public function getCollectionValueType(): self
+    {
+        if (!$this->isCollection()) {
+            throw new LogicException(sprintf('Cannot get collection value type on "%s" type as it\'s not a collection.', (string) $this));
+        }
+
+        return new self(builtinType: self::BUILTIN_TYPE_MIXED);
+    }
+
+    public function isList(): bool
```
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: Remove

File: src/Symfony/Component/TypeInfo/Type.php [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+    private bool $isCollection;
+
+    private bool $isEnum;
+
+    private bool $isBackedEnum;
+
+    public function __construct(
+        string $builtinType = null,
+        string $className = null,
```
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: BackedEnumType

File: src/Symfony/Component/TypeInfo/Type.php [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+        if (!\in_array($builtinType, self::BUILTIN_TYPES, true)) {
+            throw new InvalidArgumentException(sprintf('"%s" is not a valid PHP type.', $builtinType));
+        }
+
+        if (self::BUILTIN_TYPE_OBJECT === $builtinType && null === $className) {
+            $className = \stdClass::class;
+        }
+
+        if (is_subclass_of($className, \UnitEnum::class)) {
```
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: UnitEnumType

File: src/Symfony/Component/TypeInfo/Type.php [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    {
+        return $this->builtinType;
+    }
+
+    public function isNullable(): bool
+    {
+        return \in_array($this->builtinType, [self::BUILTIN_TYPE_NULL, self::BUILTIN_TYPE_MIXED], true);
+    }
+
+    public function isScalar(): bool
```
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: isBuiltin

File: src/Symfony/Component/TypeInfo/Type.php [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+namespace Symfony\Component\TypeInfo;
+
+use Symfony\Component\TypeInfo\Exception\InvalidArgumentException;
+use Symfony\Component\TypeInfo\Exception\LogicException;
+
+/**
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ * @author Baptiste Leduc <baptiste.leduc@gmail.com>
+ */
+readonly class Type implements \Stringable
```
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: ?extend \ReflectionNamedType?

File: src/Symfony/Component/TypeInfo/Type.php [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            self::BUILTIN_TYPE_FLOAT,
+            self::BUILTIN_TYPE_STRING,
+            self::BUILTIN_TYPE_BOOL,
+            self::BUILTIN_TYPE_NULL,
+            self::BUILTIN_TYPE_FALSE,
+            self::BUILTIN_TYPE_TRUE,
+        ], true);
+    }
+
+    public function isObject(): bool
```
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: Remove

File: src/Symfony/Component/TypeInfo/Type.php (hunk 3 above)
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: ObjectType

File: src/Symfony/Component/TypeInfo/Type.php (hunk 6 above)
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: BuiltinType

File: src/Symfony/Component/TypeInfo/Type.php [hunk 8]
```diff
@@ -48,4 +53,30 @@
...
+    {
+        if (!in_array($builtinType, self::BUILTIN_TYPES)) {
+            throw new InvalidArgumentException(sprintf('"%s" is not a valid PHP type.', $builtinType));
+        }
+
+        if ($this instanceof BuiltinType && $builtinType === $this->getBuiltinType()) {
+            return true;
+        }
+
+        if ($this instanceof UnionType || $this instanceof IntersectionType) {
```
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: I don't think that a builtin can be part of an intersection type

File: src/Symfony/Component/TypeInfo/Type.php [hunk 9]
```diff
@@ -48,4 +53,30 @@
         self::BUILTIN_TYPE_ITERABLE,
         self::BUILTIN_TYPE_MIXED,
     ];
+
+    public function isType(string $builtinType): bool
```
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: `isBuiltinType`?

File: src/Symfony/Component/TypeInfo/Type.php [hunk 10]
```diff
@@ -48,4 +53,30 @@
...
+                }
+            }
+        }
+
+        return false;
+    }
+
+    public function isNullable(): bool
+    {
+        return $this->isType(self::BUILTIN_TYPE_NULL);
```
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: `mixed` can be considered as `null` as well

File: src/Symfony/Component/TypeInfo/Type.php [hunk 11]
```diff
@@ -48,4 +53,30 @@
...
         self::BUILTIN_TYPE_MIXED,
     ];
+
+    public function isType(string $builtinType): bool
+    {
+        if (!in_array($builtinType, self::BUILTIN_TYPES)) {
+            throw new InvalidArgumentException(sprintf('"%s" is not a valid PHP type.', $builtinType));
+        }
+
+        if ($this instanceof BuiltinType && $builtinType === $this->getBuiltinType()) {
```
Comment (mtarld) [inline src/Symfony/Component/TypeInfo/Type.php]: What about `mixed`? It kind of contains the builtin type as well

File: src/Symfony/Component/TypeInfo/BuiltinType.php [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace Symfony\Component\TypeInfo;
+
+/**
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ * @author Baptiste Leduc <baptiste.leduc@gmail.com>
+ */
+enum BuiltinType: string
+{
+    case Array = 'array';
```
Comment (Korbeil) [inline src/Symfony/Component/TypeInfo/BuiltinType.php]: I'm not sure if we should put the enum cases uppercase or not. Is there a way to do enums within Symfony ?

File: src/Symfony/Component/TypeInfo/Type/BackedEnumType.php [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+namespace Symfony\Component\TypeInfo\Type;
+
+use Symfony\Component\TypeInfo\BuiltinType as BuiltinTypeEnum;
+use Symfony\Component\TypeInfo\Exception\InvalidArgumentException;
+
+/**
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ * @author Baptiste Leduc <baptiste.leduc@gmail.com>
+ */
+class BackedEnumType extends EnumType
```
Comment (Korbeil) [inline src/Symfony/Component/TypeInfo/Type/BackedEnumType.php]: IMO, we should declare final classes (the ones that are not extended) as final.

File: src/Symfony/Component/TypeInfo/TypeResolver/ReflectionTypeResolver.php [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        }
+
+        $builtinTypeOrClass = $subject->getName();
+        $nullable = $subject->allowsNull();
+
+        if (\in_array($builtinTypeOrClass, self::UNSUPPORTED_BUILTIN_TYPES, true)) {
+            throw new UnsupportedException(sprintf('"%s" type is not supported.', $builtinTypeOrClass));
+        }
+
+        if ('array' === $builtinTypeOrClass) {
```
Comment (Korbeil) [inline src/Symfony/Component/TypeInfo/TypeResolver/ReflectionTypeResolver.php]: Use BuiltinType enum values ?

**Review by Korbeil — APPROVED** [review]
Seems very good :tada: 

I would like to add a `isBuiltinType()` test on all type classes.

**Pull request closed without merging**